### ISSUE TITLE
Implement a URI generator that produces shortened URIs

### ISSUE THREAD

**Issue by ymark** [isl/x3ml#115]
Implement an instance generator (or update an existing one) that will produce shortened version of lengthy URIs. For example instead of creating the URI

```
http://example.org/Stiftung%20Preussischer%20Kulturbesitz
```

it will create the shortned URI:

```
http://example.org/8A6GHE
```

**Comment by ymark** [comment]
The aforementioned functionality has been implemented. For the time-being the URIs can be shortened using the built-in custom generators. The only addition is to include the shorten attribute in the generator. The following example shows the mappings and the generator policy should be defined.

The generator-policy:

```
<generator name="LocalTermURIShorten" prefix="abc" shorten="yes">
	<pattern>{hierarchy}/{term}</pattern>	
</generator>
```

The definition of the instance generator within mappings:

```
<instance_generator name="LocalTermURIShorten">
	<arg name="hierarchy" type="constant">coin</arg>
	<arg name="term" type="xpath">text()</arg>
</instance_generator>
```

This will generate URIs of the form: 

```
http://www.example.com/urek7ffq7sc4"
```

The functionality will be made available from version 1.8.4

**Comment by ymark** [comment]
Hi @alessiabardi, the hashed URIs feature has been implemented and included in version 1.8.4 as well as in the subsequent ones (1.8.5, 1.8.5-1, 1.8.6-SNAPSHOT)

**Comment by alessiabardi** [comment]
Perfect, thank you!

**Comment by Habennin** [comment]
Hi @ymark, question to you. Reading the above, I have a few use questions.

a) can I add the attribute 'shorten=yes' to custom generators or does it only work for the one new generator type called 'LocalTermURIshorten'?

I ask because we have dozens of custom generators in Parthenos and it would be a much more future proof solution if the attribute can be applied to any generator and not just a particular single function generator

b) I notice in the example that if the generator had a hierarchy term that this gets hashed as much as the actual ID term. Is this by design/necessary or is it simply a choice?

I ask because in Parthenos we have dozens of custom generators that build a list of hierarchical elements into the URI and it would be nice to retain these such that the end result would be http://example.com/hierarchicalelement/hashedstuffbasedonthevariablepassedgoeshere

To test a, I tried to update our Parthenos generators file and run it on an existing mapping in the D4Science environment, but there were no apparent changes (neither an error was thrown nor did the behaviour change). 

Let me know!

Cheers

George

**Comment by Habennin** [comment]
Ok, so I realized that the Parthenos instance of 3M is not yet at the engine version that would enable running this. So I downloaded a mapping from Parthenos and uploaded it in the Forth instance of 3M and ran it. The shorten function does run on my custom generators (therefore seemingly answering 'a' as posed above). The result, however, is a bit weird and non-functional. Instead of generating a URI within pointy  brackets a la "<UUID:AAAA>" is generates this (for example): "parthenos:rhb2fhy2mywp". The first, and crucial thing to note is that there are no pointy brackets. So the end result is not a usable RDF file (to my knowledge, anyhow it certainly breaks the RDF Viewer). Is this a bug or because I'm doing it wrong? I attach the necessary files to test this which are a zip and also the original xml source input. When I imported to 3M Forth instance this zip file the source xml input corrupted or was not transferred (have no idea which) and so I had to manually re-enter it. But that is another issue I'm guessing...

Actually I can't attach the zip as it is rejected by github so I guess I will email it to you. Otherwise it is mapping 1156 in the FORTH 3M instance and 464 in the Parthenos 3M instance.

**Comment by ymark** [comment]
Hi @Habennin . As regards your questions

1) The hashed functionality can be applied for all the custom generators. LocalTermURIShorten is just an identifier for a generator I created. I saw in the provided X3ML files that you are using it with the **ThingURI** generator. So that should work.

2) It was just a choice, Leaving the hierarchy out of the hash function could still lead to lengthy (yet readable) URIs. Of course, if there is such a requirement then we can consider it.

As regards the issues you mention, I'll check them out.

**Comment by ymark** [comment]
We've updated the URI shortening functionality. In particular the major updates are the following: 

* instead of creating a custom hash (e.g. rhb2fhy2mywp) it exports a UUID-like hash (e.g. 3644A684-F98E-38FE-A23C-713B77189A77)
* It now hashes only the last part of the URI. This allows preserving parts of the URI that reveal the hierarchy (e.g. The URI http://localhost/object/appellation/the%20name%20of%20the%object will be hashed to http://localhost/coin/appellation/802DE79F-E672-3EDE-9A69-0C64B5A391E0). 

The functionality will be made available from version 1.9.2
